### DISCUSSION TITLE
Button in facet "actions" of accordionPanel not working if tab is inactive

### DISCUSSION THREAD

**Discussion by FranzHagel** [primefaces/community#2206] [PrimeFaces] [answered]
The reason is in method visitTree of UITabPanel: Only when the tab.isLoaded() it visits the rest of the tab.

**Comment by vieira-brz** [comment]
The issue you're encountering with the PrimeFaces `AccordionPanel` where buttons within the "actions" facet of an inactive tab aren't working stems from the component's behavior regarding how it handles the rendering and processing of inactive tabs. The `AccordionPanel` component, like many tabbed UI components, optimizes performance by not fully processing children of inactive tabs unless they are specifically marked to do so. This can prevent actions within these tabs from executing as they have not been fully initialized or rendered in the DOM.

### Understanding the Problem

In PrimeFaces, the `visitTree` method is responsible for iterating over the component tree. For `UITabPanel` (which `AccordionPanel` extends from), this method checks if a tab is loaded (`tab.isLoaded()`) before visiting the components within that tab. If the tab is not loaded, the components inside it are not processed, and any listeners attached to those components (like buttons) are not active.

### Workaround: Enforcing Tab Loading

One common workaround is to ensure that the tabs are loaded when the `AccordionPanel` is instantiated. This can be achieved by setting the `dynamic` attribute of the `AccordionPanel` to `false`, which forces all tabs to load. However, this approach might increase the initial load time since all the contents of all tabs are rendered at once.

```xml
<p:accordionPanel dynamic="false">
    <!-- Your tabs and contents here -->
</p:accordionPanel>
```

### Better Solution: Lazy Loading with Pre-Activation

If performance is a concern and you cannot load all tabs at once, you can programmatically ensure that the tab containing the action buttons is loaded when needed. This can be done by using a little JavaScript to activate the tab before the action is triggered, or by backend methods to manage the state of the tabs.

#### JavaScript Approach:

You can use JavaScript to open the tab before executing the action. For example, you can bind a click event to the button that first switches to the tab and then performs the button's action.

```javascript
PF('widgetVarForAccordion').select(tabIndex); // Select the tab first
// Then perform the button action
```

#### Backend Approach:

Manage which tab should be active on the server side, and ensure the tab is loaded when it needs to interact with components within it.

```java
// Example in a managed bean method
public void onButtonAction() {
    // Set the active tab index, ensure it loads
    accordionPanel.setActiveIndex(requiredTabIndex);
    // Logic for the action
}
```

### PrimeFaces Configuration

Additionally, check your `web.xml` or any relevant configuration to ensure that the state saving method is appropriate and does not interfere with dynamic behaviors:

```xml
<context-param>
    <param-name>javax.faces.PARTIAL_STATE_SAVING</param-name>
    <param-value>true</param-value>
</context-param>
```

**Reply by FranzHagel** [reply]
The buttons purpose is to hide the panel and show a panel with input fields. So opening the panel before hiding it is not optimal.
BTW, the backing bean is not called. This is my core problem.
Now I try a custom renderer with a visitCallback in its decode() method.

**Reply by vieira-brz** [reply]
I got it, so try this:

```java
public class CustomAccordionPanelRenderer extends AccordionPanelRenderer {

    @Override
    public void decode(FacesContext context, UIComponent component) {
        if (component instanceof UITabPanel) {
            UITabPanel tabPanel = (UITabPanel) component;
            for (UITab tab : tabPanel.getTabs()) {
                if (!tab.isLoaded()) {
                    tab.setLoaded(true); // Ensure tab is loaded for visiting
                }
            }
        }
        
        // Visit the tree to ensure all components are processed
        component.visitTree(VisitContext.createVisitContext(context), (visitContext, visitNode) -> {
            visitNode.getComponent().decode(context);
            return VisitResult.ACCEPT;
        });

        super.decode(context, component);
    }
}
```

Register this custom renderer in your `faces-config.xml`:

```xml
<render-kit>
    <renderer>
        <component-family>org.primefaces.component</component-family>
        <renderer-type>org.primefaces.component.AccordionPanelRenderer</renderer-type>
        <renderer-class>com.yourpackage.CustomAccordionPanelRenderer</renderer-class>
    </renderer>
</render-kit>
```

**Reply by FranzHagel** [reply]
It does not get to the decode() method of the Renderer, because the UITabPanel.visitTree() does not call visitTree, when the Tab is not loaded. I was on the wrong path with this try.

**Reply by vieira-brz** [reply, marked as answer]
And this?

```java
public class CustomUITabPanel extends UITabPanel {

    @Override
    public boolean visitTree(VisitContext context, VisitCallback callback) {
        // Temporarily force all tabs to be considered loaded
        List<UITab> originalLoadedTabs = new ArrayList<>();
        for (UITab tab : getTabs()) {
            if (!tab.isLoaded()) {
                tab.setLoaded(true);
                originalLoadedTabs.add(tab);  // Keep track of tabs that were not originally loaded
            }
        }

        try {
            // Visit the tree normally now that all tabs are marked as loaded
            return super.visitTree(context, callback);
        } finally {
            // Restore the original 'loaded' state to tabs that were not loaded before
            for (UITab tab : originalLoadedTabs) {
                tab.setLoaded(false);
            }
        }
    }
}
```

In your `faces-config.xml`:

```xml
<component>
    <component-type>org.primefaces.component.UITabPanel</component-type>
    <component-class>com.yourpackage.CustomUITabPanel</component-class>
</component>
```

**Reply by FranzHagel** [reply]
Thank you very much! This works!
I overrided AccordionPanel, since I don' t want it in TabView.
